src/Bicep.LangServer/Registry/ModuleRestoreScheduler.cs

```diff
@@ -0,0 +1,134 @@
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Bicep.LanguageServer.Registry
+{
+    public sealed class ModuleRestoreScheduler : IModuleRestoreScheduler, IDisposable
+    {
+        private record QueueItem(ICompilationManager CompilationManager, DocumentUri Uri, ImmutableArray<ModuleDeclarationSyntax> References);
+
+        private record CompletionNotification(ICompilationManager CompilationManager, DocumentUri Uri);
+
+        private readonly IModuleDispatcher moduleDispatcher;
+
+        private readonly Queue<QueueItem> queue = new();
+
+        private readonly CancellationTokenSource cancellationTokenSource = new();
+
+        // block on initial wait until signaled
+        private readonly ManualResetEventSlim manualResetEvent = new(false);
+
+        private bool disposed = false;
+        private Task? consumerTask;
+
+        public ModuleRestoreScheduler(IModuleDispatcher moduleDispatcher)
+        {
+            this.moduleDispatcher = moduleDispatcher;
+        }
+
+        /// <summary>
+        /// Requests that the specified modules be restored to the local file system.
+        /// Does not wait for the operation to complete and returns immediately.
+        /// </summary>
+        /// <param name="references">The module references</param>
+        public void RequestModuleRestore(ICompilationManager compilationManager, DocumentUri documentUri, IEnumerable<ModuleDeclarationSyntax> references)
+        {
+            this.CheckDisposed();
+            var item = new QueueItem(compilationManager, documentUri, references.ToImmutableArray());
+            lock (this.queue)
+            {
+                this.queue.Enqueue(item);
+
+                // notify consumer about new items
+                this.manualResetEvent.Set();
+            }
+        }
+
+        public void Start()
+        {
+            this.CheckDisposed();
+            this.consumerTask = Task.Factory.StartNew(this.ProcessQueueItems, this.cancellationTokenSource.Token, TaskCreationOptions.LongRunning, TaskScheduler.Default);
+        }
+
+        public void Dispose()
+        {
+            // this is a sealed class - no need for full IDisposable implementation
+            if(!this.disposed)
+            {
+                this.disposed = true;
+                if (this.consumerTask is not null)
+                {
+                    this.cancellationTokenSource.Cancel();
+                    this.consumerTask = null;
+                }
+            }
+        }
+
+        private void ProcessQueueItems()
+        {
+            var token = this.cancellationTokenSource.Token;
+            while (true)
+            {
+                this.manualResetEvent.Wait(token);
+
+                var notifications = new HashSet<CompletionNotification>();
+                var references = new List<ModuleDeclarationSyntax>();
+                lock (this.queue)
```

Comment:
Is there a reason to use a custom locking mechanism instead of `ConcurrentQueue` here?

Reply:
Needed to make sure that the manual reset event calls don't happen in the wrong order. Throughput isn't a concern here, so it should be ok.